File: scripts/algorithms/M/Maximize Number of Subsequences in a String/Maximize Number of Subsequences in a String.js
// Runtime: 366 ms (Top 17.65%) | Memory: 69.2 MB (Top 29.41%)
/**
 * @param {string} text
 * @param {string} pattern
 * @return {number}
 */
var maximumSubsequenceCount = function(text, pattern) {
    const arrText = text.split("")
    const lengthP0 = arrText.filter(x => x === pattern[0]).length
    const lengthP1 = arrText.filter(x => x === pattern[1]).length
    const [c1, c2, lengthmax] = lengthP0 <= lengthP1 ? [...pattern, lengthP1]: [pattern[1], pattern[0], lengthP0]
    let newText = lengthP0 <= lengthP1 ? [c1,...arrText]: [...arrText, c1]
    newText = lengthP0 <= lengthP1 ? newText : newText.reverse()

    let i = 0;
    let count = 0;
    let countmax = lengthmax
    while(i < newText.length) {
        if(newText[i] === c1) {
            count += countmax
        }
        if(newText[i] === c2) {
            countmax--
        }
        i++;
    }
    return count
};

File: scripts/algorithms/M/Maximize Number of Subsequences in a String/Maximize Number of Subsequences in a String.java
// Runtime: 83 ms (Top 15.87%) | Memory: 54.4 MB (Top 56.08%)
class Solution {
    public long maximumSubsequenceCount(String text, String pattern) {
        //when pattern[0] == pattern[1]
        if (pattern.charAt(0) == pattern.charAt(1)) {
            long freq = 1;
            //O(N)
            for (int i = 0; i < text.length(); i++) {
                if (text.charAt(i) == pattern.charAt(0)) {
                    freq++;
                }
            }
            //number of subsequences : choose any two characters from freq nC2
            return (freq * (freq - 1)) / 2;
        }

        //choice 1
        String text1 = pattern.charAt(0) + text;

        int freq = 0;
        long count1 = 0;
        //O(N)
        for (int i = 0; i < text1.length(); i++) {
            if (text1.charAt(i) == pattern.charAt(0)) {
                freq++;
            } else if (text1.charAt(i) == pattern.charAt(1)) {
                count1 += freq;
            }
        }

        //choice 2
        String text2 = text + pattern.charAt(1);
        freq = 0;
        long count2 = 0;
        //O(N)
        for (int i = text2.length() - 1; i>= 0; i--) {
            if (text2.charAt(i) == pattern.charAt(1)) {
                freq++;
            } else if (text2.charAt(i) == pattern.charAt(0)) {
                count2 += freq;
            }
        }

        return Math.max(count1, count2);
    }
}